### ISSUE TITLE
Cannot parse FITS unit

### ISSUE THREAD

**Issue by olebole** [astropy/astropy#11851]
### Description

I want to parse the unit "e" as u.electron from a FITS table. I am not sure whether I did it right; I however followed the message of the `ValueError` below:
```
import astropy.units as u
e = u.def_unit('e', u.electron, format={'fits': 'e'})
u.add_enabled_units(e)

fits_e = u.Unit('el, format="fits")
```

### Expected behavior

I would expect that `fits_e = e`, and that I can now open a table that has `TUNITn = 'e       '` with `QTable.read()` and get a column that has a unit convertive to `u.electron`.

### Actual behavior

I get the `ValueError`:

> 'e' did not parse as fits unit: At col 0, Unit 'e' not supported by the FITS standard.  If this is meant to be a custom unit, define it with 'u.def_unit'. To have it recognized inside a file reader or other code, enable it with 'u.add_enabled_units'. For details, see https://docs.astropy.org/en/latest/units/combining_and_defining.html

With `parse_strict="warn"`, this is a warning only, but then the created unit is an `UnrecognizedUnit`. The same happens when I try to open a table now.

### System Details

That is astropy 4.2 on Debian Testing (Bullseye).

**Comment by mhvk** [comment]
I think this may partially be a problem of the "helpful hint" part of the error message being both too general (this works for `format='generic'`) and in the wrong place: I thought it would work for tables, where the unit parsing is special-cased --- though I see that in #9015 I only ensured this worked for tables also written by astropy.

Also related: #11842 - the idea there is to allow fall-back to `generic` when parsing FITS units - which would mean the enabled user unit would actually be used.

**Comment by olebole** [comment]
Is there actually a way to parse non-standard units in a FITS table into a usable astropy unit?

**Comment by mhvk** [comment]
I think for a Table not produced by astropy you need to do post-processing, for each `UnrecognizedUnit` seeing if `u.Unit(str(unit))` is helpful... The idea of #11842 is to make that an explicit option.

**Comment by saimn** [comment]
> Also related: #11842 - the idea there is to allow fall-back to generic when parsing FITS units - which would mean the enabled user unit would actually be used.

Yeah I think the fallback to generic would be nice to have. In `CCDData`'s FITS reader only the generic unit is used, I guess mostly because electron is not a standard FITS unit, which is quite annoying.

**Comment by mhvk** [comment]
Do pass on your annoyances to the IAU FITS working group... https://fits.gsfc.nasa.gov/iaufwg/iaufwg.html (though I will readily admit that I'm not 100% sure astropy is up to date for FITS units - I could not find a list of supported units on https://fits.gsfc.nasa.gov/)

**Comment by olebole** [comment]
I am not sure about the use of the "generic" parser: I would guess that sometimes names are not unique for a unit, but depend on the context. For example `e` could mean either a count or a charge.
Also, usually one wouldn't want to allow just "any" unit to be read, but a certain selection on top of the FITS standard:
Some people would just want to add `'flam'` to the list of units, or 'electron' as a count.
Therefore, I would (instead of just using the "generic" reader) rather like to have the FITS unit parser extensible, by the API similar to the example in the top posting (since this is what one would usually expect).
Another advantage would be that one would not just switch to "lazy", but still keep a "standard, with well-defined exceptions".

**Comment by mhvk** [comment]
Good points. Actually, what perhaps would be easiest is if the new unit aliases were understood by all unit readers. Temporary aliases is more or less what they are meant for:
https://docs.astropy.org/en/latest/api/astropy.units.add_enabled_aliases.html#astropy.units.add_enabled_aliases
